staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -193,6 +200,15 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationapi.APIServic
 		return
 	}
 
+	if apiregistrationapi.IsAPIServiceConditionFalse(apiService, apiregistrationapi.Available) {
+		r.handlingInfo.Store(proxyHandlingInfo{
+			serviceName:             apiService.Spec.Service.Name,
+			serviceNamespace:        apiService.Spec.Service.Namespace,
+			serviceUnAvailableError: fmt.Errorf("service %s/%s is unavailable", apiService.Spec.Service.Namespace, apiService.Spec.Service.Name),
```

Comment:
storing an error in the struct is a bit odd. Why not creating the error object in-place at https://github.com/kubernetes/kubernetes/pull/58070/files/f6cb611c66079468b10b67daacaf2e019e407657..1cbe752ab44b2314f66fbbefa255222efcd31ba7#diff-7e29367710c98f4dbb7b727306e3a2a0R98 ?

Reply:
```
func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
```

`proxyHandler` is a `http.Handler`, can't get a `APIService` from this function.

And there is already a similar attribute `transportBuildingError `: https://github.com/weekface/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go#L70

Reply:
but that one seems to be optional. serviceUnAvailableError is not (should not be).

Reply:
`serviceUnAvailableError` is also optional, only set when the service is unavailable.

Reply:
Do you mean that
```
serviceUnAvailableError error
```
 should be changed to:
```
serviceUnAvailable bool
```
?

Reply:
yes, the bool looks better.

Reply:
Or changed to
```
serviceAvailable bool
``

Then, default, service is unavailable. 

We set this attribute to `true` only when the `apiservice controller` synced and find the service is available?

Reply:
sounds good